docs/features/async-streams.md

```diff
@@ -274,39 +278,46 @@ finallyEntryLabel:
 }
 ```
 
-In both cases, we will add a `if (disposeMode) /* jump to next finally or exit */` after the block for `finally` logic.
+In both cases, we will add a `if (disposeMode) /* jump to enclosing finally or exit */` after the block for `finally` logic.
 
 #### State values and transitions
 
 The enumerable starts with state -2.
-Calling GetAsyncEnumerator sets the state to -1, or returns a fresh enumerator (also with state -1).
+Calling GetAsyncEnumerator sets the state to -3, or returns a fresh enumerator (also with state -3).
 
 From there, MoveNext will either:
-- reach the end of the method (-2)
-- reach a `yield break` (-1, dispose mode = true)
-- reach a `yield return` or `await` (N)
+- reach the end of the method (-2, we're done and disposed)
+- reach a `yield break` (state unchanged, dispose mode = true)
+- reach a `yield return` (-N, decreasing from -4)
+- reach an `await` (N, increasing from 0)
 
-From suspended state N, MoveNext will resume execution (-1).
-But if the suspension was a `yield return`, you could also call DisposeAsync, which resumes execution (-1) in dispose mode.
+From suspended state N or -N, MoveNext will resume execution (-1).
+But if the suspension was a `yield return` (-N), you could also call DisposeAsync, which resumes execution (-1) in dispose mode.
 
 When in dispose mode, MoveNext continues to suspend (N) and resume (-1) until the end of the method is reached (-2).
 
+The result of invoking `DisposeAsync` from states -1 or N is unspecified. This compiler throws `NotSupportException` for those cases.
```

Comment:
>This compiler throws [](start = 74, length = 20)

Minor point: Perhaps "The compiler generates `throw NotSupportedException` ..." or "The generated code throws `NotSupportedException` ..."

Reply:
Definitely. I'll include this fix in my next async-streams PR.

---
In reply to: [245376450](https://github.com/dotnet/roslyn/pull/31764#discussion_r245376450) [](ancestors = 245376450)